### PR TITLE
rework "journalctl -M"

### PR DESCRIPTION
**Pull request by poettering**
Let's make sure "journalctl -M" works on containers whose full FS is not visible on the host, and other fixes.

### CODE REVIEW

**Comment by evverx** [comment]
I run a container without `systemd`:

```
# mkdir -p cont/bin cont/sbin cont/etc
# touch cont/etc/machine-id
# cp $(type -P busybox) cont/bin
# cp /etc/os-release cont/etc/
# ln -s busybox cont/bin/sh
# ln -s ../bin/busybox cont/sbin/init
# systemd-nspawn -D cont/ -b &

# machinectl --no-pager --no-legend
cont    container systemd-nspawn
```

```
# journalctl -M cont
-- Logs begin at Mon 2016-04-25 11:18:18 UTC, end at Mon 2016-04-25 11:25:50 UTC
Apr 25 11:18:18 systemd-testsuite systemd-journald[389]: Runtime journal (/run/l
Apr 25 11:18:18 systemd-testsuite kernel: Initializing cgroup subsys cpuset
Apr 25 11:18:18 systemd-testsuite kernel: Initializing cgroup subsys cpu
Apr 25 11:18:18 systemd-testsuite kernel: Initializing cgroup subsys cpuacct
Apr 25 11:18:18 systemd-testsuite kernel: Linux version 4.4.0-21-generic (buildd
Apr 25 11:18:18 systemd-testsuite kernel: Command line: root=/dev/sda1 systemd.l
Apr 25 11:18:18 systemd-testsuite kernel: KERNEL supported cpus:
Apr 25 11:18:18 systemd-testsuite kernel:   Intel GenuineIntel
Apr 25 11:18:18 systemd-testsuite kernel:   AMD AuthenticAMD
Apr 25 11:18:18 systemd-testsuite kernel:   Centaur CentaurHauls
```

(logs from the host)

Should be:

```
# journalctl -M cont
No journal files were found.
-- No entries --
```

**Comment by poettering** [comment]
Hmm. thats really strange. Any chance you can provide an strace of that journalctl invocation?

**Comment by evverx** [comment]
Sure: https://gist.github.com/evverx/f3c159b376c4e94a9749c088a37b478f

**Comment by poettering** [comment]
ok, thanks, found the issue

**Comment by poettering** [comment]
@evverx force pushed a new version that fixes the issue. please have a look!

File: src/systemd/sd-journal.h [hunk 1]
```diff
@@ -88,7 +88,7 @@ int sd_journal_open_directory(sd_journal **ret, const char *path, int flags);
 int sd_journal_open_directory_fd(sd_journal **ret, int fd, int flags);
 int sd_journal_open_files(sd_journal **ret, const char **paths, int flags);
 int sd_journal_open_files_fd(sd_journal **ret, int fds[], unsigned n_fds, int flags);
-int sd_journal_open_container(sd_journal **ret, const char *machine, int flags);
+int sd_journal_open_container(sd_journal **ret, const char *machine, int flags); /* deprecated */
```
Comment (keszybz) [inline src/systemd/sd-journal.h:91]: <strike>man/sd_journal_open.xml should be updated too.</strike> OK, I see it's already done down the road.

**Comment by evverx** [comment]
Works fine:

```
# journalctl -M cont
No journal files were found.
-- No entries --
```

Thanks!

But:

```
# journalctl -M cont -f
No journal files were found.
Failed to get journal fd: Bad file descriptor
```

`strace`: https://gist.github.com/evverx/57348f151d8df11b9e88216104b454d7

`systemd`-container:

```
# systemd-nspawn -D nspawn-root/ -b systemd.unit=rescue.target &

# journalctl -M nspawn-root --no-pager
-- Logs begin at Mon 2016-04-25 14:29:16 UTC, end at Mon 2016-04-25 14:29:19 UTC. --
Apr 25 14:29:16 CONTAINER systemd-journald[17]: Runtime journal (/run/log/journal/) is 3.0M, max 24.4M, 21.3M free.
Apr 25 14:29:17 CONTAINER systemd-journald[17]: System journal (/var/log/journal/) is 4.6M, max 36.9M, 32.3M free.
Apr 25 14:29:17 CONTAINER systemd-journald[17]: Time spent on flushing to /var is 17.356ms for 2 entries.
Apr 25 14:29:17 CONTAINER systemd-journald[17]: Journal started
Apr 25 14:29:17 CONTAINER systemd[1]: Starting Flush Journal to Persistent Storage...
Apr 25 14:29:18 CONTAINER systemd[1]: Started Flush Journal to Persistent Storage.
Apr 25 14:29:18 CONTAINER systemd[1]: Starting Create Volatile Files and Directories...
Apr 25 14:29:18 CONTAINER systemd[1]: Started Create Volatile Files and Directories.
Apr 25 14:29:18 CONTAINER systemd[1]: Starting Update UTMP about System Boot/Shutdown...
Apr 25 14:29:18 CONTAINER systemd[1]: Reached target System Time Synchronized.
Apr 25 14:29:18 CONTAINER systemd[1]: Started Update UTMP about System Boot/Shutdown.
Apr 25 14:29:18 CONTAINER systemd[1]: Reached target System Initialization.
Apr 25 14:29:18 CONTAINER systemd[1]: Started Rescue Shell.
Apr 25 14:29:18 CONTAINER systemd[1]: Reached target Rescue Mode.
Apr 25 14:29:18 CONTAINER systemd[1]: Starting Update UTMP about System Runlevel Changes...
Apr 25 14:29:19 CONTAINER systemd[1]: Started Update UTMP about System Runlevel Changes.
Apr 25 14:29:19 CONTAINER systemd[1]: Startup finished in 4.781s.

# journalctl -M nspawn-root --no-pager -f
Failed to get journal fd: Bad file descriptor
```

`strace`: https://gist.github.com/evverx/425c14cdf4e8347f8d8024a3858d3391

**Comment by keszybz** [comment]
Looks all good, apart from the issue with -f, I'm seeing that too.

**Comment by poettering** [comment]
Ah, indeed, i borked that. Pushed a new version, that fixes this. Please have a look! thanks!

**Comment by evverx** [comment]
Seems like `-f` doesn't properly handle reboots:

```
# machinectl reboot nspawn_root & journalctl -M nspawn_root -f
[1] 963
-- Logs begin at Mon 2016-04-25 16:24:37 UTC. --
Apr 25 16:34:07 systemd-testsuite systemd-resolved[25]: Using system hostname 'systemd-testsuite'.
Apr 25 16:34:07 systemd-testsuite dbus[27]: [system] Successfully activated service 'org.freedesktop.systemd1'
Apr 25 16:34:07 systemd-testsuite systemd[1]: Started Network Name Resolution.
Apr 25 16:34:07 systemd-testsuite systemd[1]: Started Login Service.
Apr 25 16:34:07 systemd-testsuite systemd-logind[22]: New seat seat0.
Apr 25 16:34:07 systemd-testsuite systemd[1]: Reached target Multi-User System.
Apr 25 16:34:07 systemd-testsuite systemd[1]: Starting Update UTMP about System Runlevel Changes...
Apr 25 16:34:07 systemd-testsuite systemd[1]: Started Update UTMP about System Runlevel Changes.
Apr 25 16:34:08 systemd-testsuite systemd[1]: Startup finished in 6.945s.
Apr 25 16:35:03 systemd-testsuite systemd[1]: Received SIGINT.
...HANGS...
^C

# journalctl -M nspawn_root -f
-- Logs begin at Mon 2016-04-25 16:24:37 UTC. --
Apr 25 16:39:47 systemd-testsuite systemd-resolved[22]: Negative trust anchors: 10.in-addr.arpa 16.172.in-addr.arpa 17.172.in-addr.arpa 18.172.in-addr.arpa 19.172.in-addr.arpa 20.172.in-addr.arpa 21.172.in-addr.arpa 22.172.in-addr.arpa 23.172.in-addr.arpa 24.172.in-addr.arpa 25.172.in-addr.arpa 26.172.in-addr.arpa 27.172.in-addr.arpa 28.172.in-addr.arpa 29.172.in-addr.arpa 30.172.in-addr.arpa 31.172.in-addr.arpa 168.192.in-addr.arpa corp home internal intranet lan local private test
Apr 25 16:39:47 systemd-testsuite systemd-resolved[22]: Using system hostname 'systemd-testsuite'.
Apr 25 16:39:48 systemd-testsuite dbus[27]: [system] Successfully activated service 'org.freedesktop.systemd1'
Apr 25 16:39:48 systemd-testsuite systemd[1]: Started Network Name Resolution.
Apr 25 16:39:48 systemd-testsuite systemd[1]: Started Login Service.
Apr 25 16:39:48 systemd-testsuite systemd-logind[25]: New seat seat0.
Apr 25 16:39:48 systemd-testsuite systemd[1]: Reached target Multi-User System.
Apr 25 16:39:48 systemd-testsuite systemd[1]: Starting Update UTMP about System Runlevel Changes...
```

Should be:

```
Apr 25 18:11:26 systemd-testsuite systemd[1]: Stopped target Local File Systems (Pre).
Apr 25 18:11:26 systemd-testsuite systemd[1]: Reached target Unmount All Filesystems.
Apr 25 18:11:26 systemd-testsuite systemd[1]: Reached target Final Step.
Apr 25 18:11:26 systemd-testsuite systemd[1]: Starting Reboot...
Apr 25 18:11:26 systemd-testsuite systemd[1]: Stopped target Swap.
Apr 25 18:11:26 systemd-testsuite systemd[1]: Shutting down.
Apr 25 18:11:26 systemd-testsuite systemd-journald[16]: Journal stopped
-- Reboot --
Apr 25 18:11:26 systemd-testsuite systemd-journald[13]: Runtime journal (/run/log/journal/) is 8.0M, max 100.0M, 92.0M free.
Apr 25 18:11:26 systemd-testsuite systemd-journald[13]: System journal (/var/log/journal/) is 8.0M, max 986.1M, 978.1M free.
Apr 25 18:11:26 systemd-testsuite systemd-journald[13]: Time spent on flushing to /var is 5.085ms for 2 entries.
Apr 25 18:11:26 systemd-testsuite systemd-journald[13]: Journal started
Apr 25 18:11:26 systemd-testsuite systemd[1]: Mounted FUSE Control File System.
Apr 25 18:11:26 systemd-testsuite systemd[1]: Starting Flush Journal to Persistent Storage...
Apr 25 18:11:26 systemd-testsuite systemd[1]: Started Load/Save Random Seed.
Apr 25 18:11:26 systemd-testsuite systemd[1]: Started Flush Journal to Persistent Storage.
Apr 25 18:11:26 systemd-testsuite systemd[1]: Starting Create Volatile Files and Directories...
Apr 25 18:11:26 systemd-testsuite systemd[1]: Started Create Volatile Files and Directories.
Apr 25 18:11:26 systemd-testsuite systemd[1]: Reached target System Time Synchronized.
Apr 25 18:11:26 systemd-testsuite systemd[1]: Starting Update UTMP about System Boot/Shutdown...
```

**Comment by poettering** [comment]
Urks, yeah, not sure we can make that work... The fds used reference the old namespace, and when the container dies that becomes useless cruft. 

Maybe we can watch the container and terminate when it goes down. But I figure that should be a separate PR. And not even sure how we could implement that at all nicely...

**Comment by evverx** [comment]
> not sure we can make that work...

That works on the current master:) So technically it's a regression.

> But I figure that should be a separate PR

I agree

**Comment by keszybz** [comment]
> >    But I figure that should be a separate PR
> 
> I agree

Me too :)

**Pull request merged**